script_based_workflows chapter

thebioengineer opened the pull request:
# Summary

Adding a chapter describing the 'official' recommendation on how `drake` will handle script-based workflows. This will address issue #41, and is in reference to the recently accepted pull request: ropensci/drake#1007

### Chapter contents
The chapter is still being fleshed out, but will have these general components:
- [x] Start with a workflow with numbered scripts, e.g. 01_data.R, 02_munge.R, 03_analysis.R, etc.
- [x] Make a function out of each script (ropensci/drake#994).
- [x] Call all the functions in a simple plan.
- [x] Iterate between drake_plan() and vis_drake_graph() to connect the targets in the correct order. 
- [x] End up with a working plan.

# Additional Chapter sections
- [x] Walk through some refactoring.
- [x] Edit each script file: put all the code in a function.
- [ ] Replace script_to_function() with source().
- [x] Decompose the functions into smaller functions with formal arguments.

# Related GitHub issues and pull requests

- Ref: #41 

# Checklist

- [x] I understand and agree to this repository's [code of conduct](https://github.com/ropensci/drake-manual/blob/master/CONDUCT.md).
- [ ] I have listed any substantial changes in the [development news](https://github.com/ropenscilabs/drake-manual/blob/master/NEWS.md).
- [ ] This pull request is not a [draft](https://github.blog/2019-02-14-introducing-draft-pull-requests).

wlandau commented on script_based_workflows.Rmd:
```diff
@@ -0,0 +1,190 @@
+
+## Building the connections
+
+If we knew what scripts need to be run before others, we can use `targets` to indicate to `drake` in which order scripts need to be evaluated. 
```
Small suggestion "we can use `targets` as function arguments to indicate..."

wlandau commented on script_based_workflows.Rmd:
```diff
@@ -0,0 +1,190 @@
+
+If we knew what scripts need to be run before others, we can use `targets` to indicate to `drake` in which order scripts need to be evaluated. 
+
+In the case described above, the final script, "03_histogram.R", generates a plot of a histogram using the data generated from "02_munge.R", which in turn relies on "01_data.R" to load the data. "01_data.R" is the first script that needs to be evaluated and does not rely on any other inputs.
```
Maybe add an explicit mention that function arguments express this dependency relationship.

wlandau commented:
Ellis, this is an excellent start! As you continue with the draft, maybe you could add

1. Some iteration: change a script and see what happens in `make()`.
2. Maybe a short conclusion mentioning:
    1. drake still works best with manually-written functions, and
    2. `Rclean` might help with refactoring.

I will likely have more comments later, but I think this keeps the ball rolling.

wlandau commented:
One thing that occurred to me: for the iteration part, we would do well to remind users to `source()` the script with `code_to_function()` before running `make()` again.

thebioengineer commented:
Hey @wlandau,

I've added a conclusion, more details on how targets need to be passed as arguments to transfer dependency, and ensured everything now runs now that we have the example from `drake_example()` running. I am still reviewing wording/making sure we covered everything we feel is important. 

Let me know if you think we've covered everything -and maybe we should ask jenny or someone on that team to review to see if it satisfies their thoughts on incorporating `drake` into script-based workflows.

wlandau commented on script_based_workflows.Rmd:
````diff
@@ -0,0 +1,201 @@
+```
+
+
+Any time a script is updated, either `functions.R` can be sourced again to update the representative function, or the specific `code_to_function()` line can be rerun. `drake` can see that the function is updated and when `make()` is called, it will rerun the updated function and all dependencies. 
````
I think a demo of this, plus a couple `vis_drake_graph()`s, may be helpful.

thebioengineer replied:
Done :)

wlandau commented on script_based_workflows.Rmd:
```diff
@@ -0,0 +1,201 @@
+By turning these scripts into functions, they can be added into a `drake` plan for workflow management.
+
+
+## Dependencies
```
People often think of dependencies as R packages and system requirements. When we talk about `drake`, and about declarative workflows in general, we use that word differently. In this context, dependencies are targets, functions, files, etc., anything that could invalidate downstream results when changed. So I think it is worth touching on what we mean by "dependency" and why it matters, at least as far as being able to skip up-to-date computation. I should probably go back earlier in the manual to make this clear in other places as well.

thebioengineer replied:
I added a little more on what a dependency is, but part of me assumes that people have leafed through the manual to describe what a `dependency` is. We can add more details if you think it would be helpful though!

wlandau commented:
Awesome! I added a couple more suggestions for minor things to expand. After that, I think we will be ready to polish the writing.

I agree that spreading the word will help, both reaching out to key people and broadcasting on Twitter and rOpenSci Slack. Maybe after this PR is merged and the HTML is rendered at ropenscilabs.github.io/drake-manual? The manual is a living document anyway, so we can still respond to feedback.

wlandau commented on _bookdown.yml:
```diff
@@ -10,6 +10,7 @@ rmd_files: [
   "walkthrough.Rmd",
   "plans.Rmd",
   "projects.Rmd",
+  "script_based_workflows.Rmd",
```
Maybe call the source file `scripts.Rmd`? I like the `{#scripts}` cross reference tag you came up with, and it would be nice for the file name to agree.

thebioengineer replied:
right, thanks for pointing that out. Ive renamed the file to scripts.Rmd. That was what I was originally planning on calling it.

wlandau commented:
I think we are ready for fine-tuning. Comments forthcoming.

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,255 @@
+dir.create("data")
+```
+
+`drake` has been built with the firm belief that the best workflows rely on approaching problems from a function perspective. This idea is supported in "Good enough practices in scientific computing", that approaching your project by writing modular code increases the end product.
````
I just realized we might want to

1. Streamline the first sentence, and
2. Talk a little about functions in plans first.

Maybe something like this:

## Function-oriented workflows

`drake` wants you to write functions for data analysis. Functions break down complicated ideas into manageable pieces.

```r
# R/functions.R
get_data <- function(file) {
  readxl::read_excel(file)
}

munge_data <- function(raw_data) {
  raw_data %>%
    mutate(Species = forcats::fct_inorder(Species))
}

fit_model <- function(munged_data) {
  lm(Sepal.Width ~ Petal.Width + Species, munged_data)
}
```

`get_data()`, `munge_data()`, and `fit_model()` are a special shorthand that makes the plan easier to read.

```r
# R/plan.R
drake_plan(
  raw_data = get_data(file_in("raw_data.xlsx")),
  munged_data = munge_data(raw_data),
  model = fit_model(munged_data)
)
```

To invoke `drake`, we load the supporting scripts and call `make()`.

```r
source("R/packages.R")
source("R/functions.R")
source("R/plan.R")
```

This function-oriented approach is elegant, powerful, testable, scalable, and maintainable. However, it can be challenging to convert pre-existing traditional script-based analyses to function-oriented `drake`-powered workflows. The stopgap in this chapter quickly retrofits `drake` and functions onto existing projects without needing to change the original underlying scripts.

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,255 @@
+
+## Traditional and Legacy Workflows
+
+A common approach to creating workflows is to write steps of the analysis into scripts that perform specific tasks. By giving these scripts meaningful names, e.g. 01_data.R, 02_munge.R, 03_analysis.R,  one can surmise both their contents and the order in which these scripts should be run.
```
Suggestion for more direct language:

It is common to express data analysis tasks in numbered scripts.

```
01_data.R
02_munge.R
03_analysis.R
```

The numeric prefixes indicate the order in which these scripts should run.

```r
# run_everything.R
source("01_data.R")
source("02_munge.R")
source("03_analysis.R")
```

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,255 @@
+
+A common approach to creating workflows is to write steps of the analysis into scripts that perform specific tasks. By giving these scripts meaningful names, e.g. 01_data.R, 02_munge.R, 03_analysis.R,  one can surmise both their contents and the order in which these scripts should be run.
+
+It was acknowledged in the (What we left out)[https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1005510#sec027] section that build tools to manage projects when code gets out of date are ideal, but can be substituted with a simple master script called `make.R` that reruns each of the scripts in order.
```
- In the link, the square brackets go around the text, and the parentheses go around the hyperlink. 
- Maybe show a Makefile with the scripts.

wlandau replied:
Hmm..., maybe an actual Makefile gets too technical.

wlandau replied:
In fact, we can probably just skip the discussion of build tools and jump to `code_to_function()`. Sketch:

`code_to_function()` can create a plan-ready function from each numbered script.

```{r, eval = TRUE}
# R/functions.R
load_data <- code_to_function("01_data.R")
do_munge  <- code_to_function("02_munge.R")
generate_histogram <- code_to_function("03_histogram.R")
do_regression <- code_to_function("04_regression.R")
generate_report <- code_to_function("05_report.R")

print(load_data)
``` 

Next, we develop a plan on top of these functions.

wlandau replied:
Maybe that last comment belongs in the section on overcoming technical debt.

wlandau commented:
Thanks to your work on the end-to-end draft, it is easier to crystallize this chapter into a flow of ideas. My latest comments are more in-depth than I expected them to be.

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,255 @@
+
+```{r}
+
+load_data <- code_to_function("01_data.R")
````
Here, it would be nice to turn each of the scripts into a function and print one of the functions.

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,255 @@
+
+## Dependencies
+
+An important aspect of drake plans is the capability to track the dependencies of each function and only rerun the parts of the plan that have updated, either in code or input. Dependencies in the terms of `drake` means to rely on prior steps to have been run and generated output that will then be used in the current step. Any time the prior step updates, then any down-stream steps that rely on i ...
```
Suggestion: "Dependencies in the terms of `drake` means to rely on prior steps to have been run and generated output that will then be used in the current step." => "In `drake`, a target's dependencies are the things it needs in order to build. Dependencies can include functions, files, and other targets upstream."

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,255 @@
+
+```{r}
+
+load_data <- code_to_function("01_data.R")
````
We can probably define these functions in the previous section and focus on the plan in this section.

thebioengineer commented:
> Thanks to your work on the end-to-end draft, it is easier to crystallize this chapter into a flow of ideas. My latest comments are more in-depth than I expected them to be.

Thanks for the in-depth feedback @wlandau. This first draft was intended to be a shot in the dark of what we might want to then mold and re-work as necessary, so glad to hear that it helped formulate the ideas. I'll work on incorporating your feedback and let you know once I think it is ready for another review!

thebioengineer commented:
@wlandau , I made updates to the text that I think captures your edits. Let me know your thoughts!

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,318 @@
+```
+## Function-oriented workflows
+
+`drake` was built around the idea that writing functions for data analysis allows
````
This chapter is really coming along nicely, so I am going to start being pickier about rhetoric.

This sentence seems a bit long and cumbersome for an intro. How about:

> `drake` works best when you write functions for data analysis. Functions break down complicated ideas into manageable pieces.

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,318 @@
+script-based analyses to function-oriented `drake`-powered workflows. This
+chapter describes a stopgap that quickly retrofits `drake` and functions onto
+existing projects without needing to change the original underlying scripts with
+minimal overhead.
```
How about:

> This chapter describes a stopgap to retrofit `drake` and functions onto
existing projects. It is quick and painless, and you do not need to change your original scripts.

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,318 @@
+
+Hopefully this chapter gave some insight on how to use the `code_to_function()`
+functionality available in `drake`. `code_to_function()`is a quick and dirty way
+to retrofit `drake` to an existing script-based project. _This approach should
```
> A large number of preexisting data science workflows consist of numbered, imperative scripts. Many such projects are already too big and cumbersome to seriously refactor. The purpose of `code_to_function()` is to quickly and painlessly allow `drake` to accommodate these situations.

However, for new projects, we strongly recommend that you write functions. Functions help organize your thoughts, and they improve portability, readability, and compatibility with `drake`. For a deeper discussion of functions and their role in `drake`, consider watching the [webinar recording of the 2019-09-23 rOpenSci Community Call](https://ropensci.org/commcalls/2019-09-24).

Even old projects are sometimes pliable enough to refactor into functions, especially with the new [`Rclean`](https://github.com/provtools/rclean) package.

thebioengineer commented on scripts.Rmd:
```diff
@@ -0,0 +1,318 @@
+simple_plan <- drake_plan(
+  data        = load_data(),
+  munged_data = munge_data(),
+  hist        = make_histogram()
```
sure, sounds like a good idea!

thebioengineer commented on scripts.Rmd:
```diff
@@ -0,0 +1,318 @@
+
+## Dependencies
+
+An important aspect of `drake` plans is the capability to track the dependencies 
```
Ah, did not mean to click on 👎 , accepted the suggestion!

thebioengineer commented:
@wlandau your suggestions have been added 💯

wlandau commented on projects.Rmd:
````diff
@@ -213,6 +217,24 @@ good_config <- drake_config(good_plan)
 vis_drake_graph(good_config, targets_only = TRUE)
 ```
 
+As of `drake` version `7.6.2.9000`, there is a function called `code_to_function()` that will assist in moving your scripts based workflow into a `drake` plan.
````
How about:

> In `drake` >= 7.6.2.9000, `code_to_function()` leverages existing imperative scripts for use in a `drake` plan.

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,258 @@
+```
+
+By turning the steps of processing into functions like `get_data()`, `munge_data()`, 
+and `fit_model()`, they become special shorthand that make the plan easier to
````
How about:

> ...and `fit_model()`, we create special shorthand to make the plan easier to read and understand.

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,258 @@
+projects. It is quick and painless, and you do not need to change your original 
+scripts.
+
+## Traditional and Legacy Workflows
```
Can we use lowercase for "Legacy" and "Workflows"? In the rest of the manual, only the first letter tends to be uppercase except for proper nouns.

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,258 @@
+
+```
+
+This looks good and dandy, but right now `drake` has no idea which part to run 
````
How about:

> It's a start, but right now...

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,258 @@
+```
+
+
+## Keeping up to date
````
> Keeping the results up to date

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,258 @@
+those targets as outdated.
+
+```{r echo=FALSE,message=FALSE}
+outdated_targets<-outdated(script_based_config)
````
Would you insert spaces around `<-`?

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,318 @@
+
+```{r echo=FALSE,message=FALSE}
+outdated_targets<-outdated(script_based_config)
+#needs to be run, for some reason `drake` is not recognizing the update when compiling
````
I just realized there is no initial `make()`! 😅  I suggest running it in a new short "Run the workflow" section between "Building the connections" and "Keeping the results up to date".

wlandau commented on scripts.Rmd:
```diff
@@ -0,0 +1,258 @@
+
+## `drake` and scripts
+
+> A large number of preexisting data science workflows consist of numbered,
```
I am not sure you need to quote yourself here.

wlandau replied:
At the beginning of the next paragraph, maybe just say "`code_to_function()` lets `drake` to accommodate script-based projects too big and cumbersome to refactor."

wlandau commented on scripts.Rmd:
````diff
@@ -0,0 +1,258 @@
+```
+
+
+## `drake` and scripts
````
Maybe call it "Conclusion" or "Final thoughts"?

thebioengineer commented:
> Ellis, I added some more comments inline. It would also be great if you would add yourself as a full author in the DESCRIPTION file and add a note to NEWS.md. After that, I will merge your PR. There may be finishing touches after that, but I think we can both work on them more casually going forward.

Great! I marked myself as an "aut", and incorporated your edits. I think this chapter reads really well now, I am curious to see how others see it!

wlandau commented:
Excellent work, Ellis! Thank you for being persistent and bearing with me. I just fixed some merge conflicts, and as soon as the CI completes, I will merge the PR.

thebioengineer commented:
Great! Happy to see that we were able to make this happen. This has been a great experience and I've learned a ton!
